File: Andromeda.AvaloniaApp.FSharp/ViewModels/AuthenticationWindowViewModel.fs
namespace Andromeda.AvaloniaApp.FSharp.ViewModels

open Andromeda.AvaloniaApp.FSharp.Helpers
open Andromeda.Core.FSharp.AppData
open Avalonia.Controls
open GogApi.DotNet.FSharp.Authentication
open ReactiveUI
open DynamicData
open System.Reactive

type AuthenticationWindowViewModel(control, parent) as this =
    inherit SubViewModelBase(control, parent)

    let code = ""
    member this.Code
        with get() = code
        and set (value: string) =
            this.RaiseAndSetIfChanged(ref code, value) |> ignore

    member val AuthenticateCommand: ReactiveCommand<Window, Unit> = ReactiveCommand.Create<Window>(this.Authenticate)

    member this.Authenticate(window: Window) =
        { this.AppData with authentication = newToken this.Code } |> this.SetAppData
        window.Close();

File: Andromeda.AvaloniaApp/ViewModels/Windows/AuthenticationWindowViewModel.cs
using Andromeda.AvaloniaApp.Helpers;
using Andromeda.AvaloniaApp.Windows;
using Avalonia.Controls;
using DynamicData;
using ReactiveUI;
using ReactiveUI.Legacy;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reactive;
using System.Text;

using static Andromeda.Core.FSharp.AppData;
using static GogApi.DotNet.FSharp.Authentication;
using static GogApi.DotNet.FSharp.Listing;

namespace Andromeda.AvaloniaApp.ViewModels.Windows
{
    public class AuthenticationWindowViewModel : ViewModelBase
    {
        private string code = "";

        public string Code
        {
            get => this.code;
            private set { this.RaiseAndSetIfChanged(ref this.code, value); }
        }

        public ReactiveCommand<Window, Unit> AuthenticateCommand { get; }

        public AuthenticationWindowViewModel(AppDataWrapper appDataWrapper) : base(appDataWrapper)
        {
            AuthenticateCommand = ReactiveCommand.Create<Window>(Authenticate);
        }

        void Authenticate(Window window)
        {
            this.SetAppData(new AppData(newToken(this.Code), this.AppData.installedGames));
            window.Close();
        }
    }
}
